src/components/ThumbnailsPanel/ThumbnailsPanel.js

```diff
@@ -55,6 +71,75 @@ class ThumbnailsPanel extends React.PureComponent {
     });
   }
 
+  onDragEnd = () => {
+    const { currentPage } = this.props;
+    const { draggingOverPageIndex, isDraggingOverTopHalf } = this.state;
+    if (draggingOverPageIndex !== null) {
+      const targetPageNumber = isDraggingOverTopHalf ? draggingOverPageIndex + 1 : draggingOverPageIndex + 2;
+
+      if (currentPage < targetPageNumber) {
+        // moving page down so target page number will decrease
+        this.afterMovePageNumber = targetPageNumber - 1;
+      } else {
+        // moving page up
+        this.afterMovePageNumber = targetPageNumber;
+      }
+
+      core.movePages([currentPage], targetPageNumber);
+    }
+
+    this.setState({ draggingOverPageIndex: null });
+  }
+
+  onPageComplete = () => {
+    if (this.afterMovePageNumber) {
+      core.setCurrentPage(this.afterMovePageNumber);
+    }
+    this.afterMovePageNumber = null;
+  }
+
+  onDragOver = (e, index) => {
+    // prevent opening pdf dropped in
+    e.preventDefault();
+    const thumbnail = e.target.getBoundingClientRect();
+
+    if (e.pageY > (thumbnail.y + thumbnail.height / 2)) {
+      this.setState({ draggingOverPageIndex: index, isDraggingOverTopHalf: false });
+    } else {
+      this.setState({ draggingOverPageIndex: index, isDraggingOverTopHalf: true });
+    }
+  }
+
+  onDragStart = (e, index) => {
+    const { currentPage } = this.props;
+
+    // need to set 'text' to empty for drag to work in FireFox and mobile
+    e.dataTransfer.setData('text', '');
+
+    if (currentPage !== (index + 1)) {
+      core.setCurrentPage(index + 1);
+    }
+  }
+
+  onDrop = e => {
+    e.preventDefault();
+    const { isThumbnailMerging } = this.props;
+    const { draggingOverPageIndex, isDraggingOverTopHalf } = this.state;
+    const { files } = e.dataTransfer;
+
+    if (isThumbnailMerging && files.length) {
+      const file = files[0];
+      const insertTo = isDraggingOverTopHalf ? draggingOverPageIndex + 1 : draggingOverPageIndex + 2;
+
+      window.readerControl.mergeDocument(file, insertTo).then(() => {
```

Comment:
Hmm in general we avoid to call readerControl within our components. The reason is that we want to make clear that functions of readerControl are for external uses.

In this case what we do is to create a new file under the helpers folder, and use that here. And in the apis folder we import the helper function to "expose" it... I know it's essentially doing the same thing, but I think we just want to remain the same mental model. Sorry

Reply:
I didn't end up adding a helper but added it to `core`